File: src/034. Digit factorials/034.rb
$factorials = [1] # 0!
1.upto(9) do |d|
  $factorials[d] = d * $factorials[d - 1]
end

def sum_of_digit_factorials(num)
  num.to_s.chars.map(&:to_i).map { |d| $factorials[d] }.inject(:+)
end

class Fixnum
  def curious?
    self == sum_of_digit_factorials(self)
  end
end

total = 0
# estimated upper limist, because 9! * 6 < 999999
upper_limit = 999_999
10.upto(upper_limit) do |n|
  total += n if n.curious?
end
puts total

File: src/034. Digit factorials/034.go
package main

import "fmt"

var factorials [10]int

func init() {
	factorials[0] = 1
	for i := 1; i <= 9; i++ {
		factorials[i] = factorials[i - 1] * i
	}
}

func main() {
	// estimated upper limist, because 9! * 6 < 999999
	const upperLimit = 999999
	sum := 0
	for i := 10; i <= upperLimit; i++ {
		if i == digitFactorial(i) {
			sum += i
		}
	}
	fmt.Println(sum)
}

func digitFactorial(n int) int {
	sum := 0
	for n > 0 {
		sum += factorials[n % 10]
		n /= 10
	}
	return sum
}
